ethereum/contracts/LightClientBridge.sol

```diff
@@ -0,0 +1,272 @@
+ */
+contract LightClientBridge {
+    using SafeMath for uint256;
+    using Bits for uint256;
+    /* Events */
+    /**
+     * @notice Notifies an observer that the prover's attempt at initital
+     * verification was successful.
+     * @dev Note that the prover must wait until `n` blocks have been mined
+     * subsequent to the generation of this event before the 2nd tx can be sent
+     * @param prover The address of the calling prover
+     * @param blockNumber The blocknumber in which the initial validation
+     * succeeded
+     * @param id An identifier to provide disambiguation
+     */
+    event InitialVerificationSuccessful(
+        address prover,
+        uint256 blockNumber,
+        uint256 id
+    );
+    /**
+     * @notice Notifies an observer that the complete verification process has
+     *  finished successfuly and the new payload will be accepted
+     * @param prover The address of the successful prover
+     * @param payload the payload which was approved for inclusion
+     * @param id the identifier used
+     */
+    event FinalVerificationSuccessful(
+        address prover,
+        bytes32 payload,
+        uint256 id
+    );
+    /* Types */
+    struct ValidationData {
+        address senderAddress;
+        bytes32 payload;
+        uint256 validatorClaimsBitfield;
+        uint256 blockNumber;
+    }
+    /* State */
+    ValidatorRegistry public validatorRegistry;
+    uint256 public currentId;
+    mapping(uint256 => ValidationData) public validationData;
+    /* Constants */
+    uint256 public constant BLOCK_WAIT_PERIOD = 45;
+    uint256 public constant MAXIMUM_NUM_SIGNERS = 167;
+
+    /**
+     * @notice Deploys the LightClientBridge contract
+     * @dev If the validatorSetRegistry should be initialised with 0 entries, then input
+     * 0x00 as validatorSetRoot
+     * @param _validatorRegistry The contract to be used as the validator registry
+     */
+    constructor(ValidatorRegistry _validatorRegistry) {
+        validatorRegistry = _validatorRegistry;
+        currentId = 0;
+    }
+
+    /* Public Functions */
+    /**
+     * @notice Executed by the prover in order to begin the process of block
+     * acceptance by the light client
+     * @param payload contains the payload signed by the validator(s)
+     * @param validatorClaimsBitfield a bitfield containing a membership status of each
+     * validator who has claimed to have signed the payload
+     * @param validatorPublicKey the public key of an arbitrary validator
+     * @param validatorSignatureCommitment the signature of an arbitrary validator
+     * @param validatorPublicKeyMerkleProof Proof required for validation of the Merkle tree
+     */
+    function newSignatureCommitment(
+        bytes32 payload,
+        uint256 validatorClaimsBitfield,
+        bytes memory validatorSignatureCommitment,
+        address validatorPublicKey,
+        bytes32[] calldata validatorPublicKeyMerkleProof
+    ) public payable {
```

Comment:
Shouldn't this be `external` instead of public? I vaguely recall `calldata` doesn't work with `public`.

Reply:
I grabbed the contracts from https://github.com/Snowfork/verification-solidity/pull/12 which is currently a WIP, just using them as a placeholder to write the core Relayer functionality against. Once the PR is merged we'll move the finalized contracts into this repo.